### DISCUSSION TITLE
`notnull` to not allow Nullable<T> too

### DISCUSSION THREAD

**Discussion by WhiteBlackGoose** [dotnet/csharplang#4904] [Language Ideas] [answered]
I'm not sure how exactly this should be designed, but the idea is to have `T` to be not null including C# 8's nullable value types. I would suggest

```cs
Quack<int?>();

public static int Quack<T>() where T : notnull => 3;
```

to be non-compilable (since `int?` is nullable), but that, I assume, would be a breaking change. However, this syntax

```cs
Quack<int?>();

public static int Quack<T>() where T : struct, notnull => 3;
```

is not compilable at all (you cannot have both `struct` and `notnull` as type constraint simultaneously), so it could be used to filter out nullables on value types.

cc @333fred

**Comment by svick** [comment, marked as answer]
Can't you already use `where T : struct` to restrict `T` to non-nullable value types?

**Reply by jnm2** [reply]
Right, `struct, notnull` would be redundant because `struct` already disallows `int?` as a generic type argument. [(sharplab)](https://sharplab.io/#v2:EYLgtghglgdgNAFxANwKYCcoDMCeAfARQFcIBjAawB5YEB+APgAoBKAbgFgAoLgAQEYAbAAIaQ4mSoAVJsyEB3ABYZUQyUJBCAzgnRFSCIQF56QgMwdOQA==)

```cs
// ❌ CS0453: The type 'int?' must be a non-nullable value type in order to use it as parameter 'T' in the generic
//  type or method 'Quack<T>()'
Quack<int?>();

static int Quack<T>() where T : struct => 3;
```

**Reply by WhiteBlackGoose** [reply]
Ohhh, this is opposite to what I think. Like, Nullable is a struct but not `notnull`, but C# assumes the exact opposite

**Reply by jnm2** [reply]
That's not quite right. C# does not allow `int?` for `notnull` without giving you a nullability warning.

```cs
#nullable enable

// ⚠ CS8714: The type 'int?' cannot be used as type parameter 'T' in the generic type or method 'Quack<T>()'.
// Nullability of type argument 'int?' doesn't match 'notnull' constraint.
Quack<int?>();

static int Quack<T>() where T : notnull => 3;
```

**Comment by ZacharyPatten** [comment]
related: https://github.com/dotnet/csharplang/discussions/707

Almost a duplicate but this issue wants constraints to interact with nullable reference syntax, which I assume would require runtime changes.

**Comment by ZacharyPatten** [comment]
Couldn't you just write an analyzer to do this though? Nowadays if you want to add usage restrictions on your code, analyzers seem like the preferred method rather than a new language feature.

You wouldn't have runtime constraint validation ~if the user was using reflection, but I doubt _"preventing users from using reflection to bypass analyzers"_ is worthy enough reason to implement a new language feature.~ Runtime constraint checking would likely be impossible here since that just isn't how nullable reference types work.

**Comment by cston** [comment]
The `notnull` constraint is part of the nullable feature and constraint violations are reported as warnings (see [nonnullable-constraint](https://github.com/dotnet/csharplang/blob/main/meetings/2019/LDM-2019-05-15.md#nonnullable-constraint)). The warnings are reported in `#nullable enable` contexts only.
